node/src/Cardano/Node/API.hs

```diff
@@ -18,33 +19,97 @@ import           Servant
+import           Pos.Core.JsonLog (CanJsonLog (..))
+import           Pos.Crypto (SecretKey)
 import qualified Pos.DB.Block as DB
 import qualified Pos.DB.BlockIndex as DB
 import qualified Pos.DB.Class as DB
 import           Pos.DB.GState.Lock (Priority (..), StateLock,
                      withStateLockNoMetrics)
-import qualified Pos.DB.Rocks.Functions as DB
-import qualified Pos.DB.Rocks.Types as DB
+import qualified Pos.DB.Rocks as DB
+import           Pos.DB.Txp (MempoolExt)
+import           Pos.DB.Txp.MemState (GenericTxpLocalData, TxpHolderTag)
 import           Pos.Infra.Diffusion.Subscription.Status (ssMap)
 import           Pos.Infra.Diffusion.Types
 import qualified Pos.Infra.Slotting.Util as Slotting
+import           Pos.Infra.Util.JsonLog.Events (jsonLogDefault)
+import           Pos.Launcher.Configuration (HasConfigurations)
 import           Pos.Launcher.Resource
 import           Pos.Node.API as Node
 import           Pos.Util (HasLens (..), HasLens')
 import           Pos.Util.CompileInfo (CompileTimeInfo, ctiGitRevision)
+import           Pos.Util.Lens (postfixLFields)
 import           Pos.Util.Servant
 import           Pos.Web (serveImpl)
+import qualified Pos.Web as Legacy
+import           Pos.WorkMode (RealModeContext)
+import           Pos.WorkMode.Class (WorkMode)
+
+type NodeV1Api
+    = "v1"
+    :> ( Node.API
+    :<|> Legacy.NodeApi
+    )
+
+nodeV1Api :: Proxy NodeV1Api
+nodeV1Api = Proxy
+
+data LegacyCtx = LegacyCtx
+    { legacyCtxTxpLocalData
+        :: !(GenericTxpLocalData ())
+    , legacyCtxPrimaryKey
+        :: !SecretKey
+    , legacyCtxUpdateConfiguration
+        :: !UpdateConfiguration
+    , legacyCtxSscContext
+        :: !SscContext
+    , legacyCtxNodeDBs
+        :: !DB.NodeDBs
+    }
+
+makeLensesWith postfixLFields ''LegacyCtx
+
+instance HasPrimaryKey LegacyCtx where
+    primaryKey = legacyCtxPrimaryKey_L
+
+instance HasLens TxpHolderTag LegacyCtx (GenericTxpLocalData ()) where
+    lensOf = legacyCtxTxpLocalData_L
+
+instance HasLens DB.NodeDBs LegacyCtx DB.NodeDBs where
+    lensOf = legacyCtxNodeDBs_L
+
+instance HasLens UpdateConfiguration LegacyCtx UpdateConfiguration where
+    lensOf = legacyCtxUpdateConfiguration_L
+
+instance HasSscContext LegacyCtx where
+    sscContext = legacyCtxSscContext_L
+
+instance {-# OVERLAPPING #-} DB.MonadDBRead (ReaderT LegacyCtx IO) where
+    dbGet = DB.dbGetDefault
+    dbIterSource = DB.dbIterSourceDefault
+    dbGetSerBlock = DB.dbGetSerBlockRealDefault
+    dbGetSerUndo = DB.dbGetSerUndoRealDefault
+    dbGetSerBlund = DB.dbGetSerBlundRealDefault
+
+legacyNodeApi :: () => LegacyCtx -> Server Legacy.NodeApi
+legacyNodeApi r =
+    hoistServer
+        (Proxy :: Proxy Legacy.NodeApi)
+        (Handler . ExceptT . try . flip runReaderT r)
+        Legacy.nodeServantHandlers
 
 launchNodeServer
-    :: NodeApiArgs
+    :: HasConfigurations
+    => NodeApiArgs
     -> NtpConfiguration
-    -> NodeResources ext
+    -> NodeResources ()
```

Comment:
It'd be nice if `NodeResources` were a `Functor` so I could `void :: Functor f => f a -> f ()` and retain polymorphism. However, the `ext` eventually bottoms out to a `TVar ext` in `GenericTxpLocalData`, and `TVar` isn't a `Functor,` so, oh well.